AP_Motors: Add multicopter thrust slew limit

Pedals2Paddles opened the pull request:
### Adds output slew rate limiting to multicopters
With this PR, ArduCopter 3.7+ will be fully compatible with all Solos, and will be included in future releases of Open Solo for all.  Other vehicle specific use cases can benefit from this as well.

_Updated 12/17/2018 MCL_

_**Brief Background**_:
Slew limiting to thrust output was a hard coded hack added to the old solo firmware to protect the ESCs from a ground bounce condition that could happen when the throttle is punched hard. It was the best that could be done when 3DR did not want to redesign the ESCs to function properly.  This was never brought over to master, which prevents Solos from being able to run modern versions of ArduCopter without hardware changes. The only way to safely run real ArduCopter versions on a Solo with slew limiting is with cumbersome hardware changes that increases ESC signal voltage to 5v instead of 3.3v.  Such as the "green cube" or logic level converters.

_**This PR**_:
New parameters `MOT_SLEW_UP_TIME` and `MOT_SLEW_DN_TIME` have been added.  The default is 0, which applies no slew limit and makes no change to existing behavior.  Valid values are 0 to 0.5 seconds.  This controls the minimum amount of time that the output can go from minimum to maximum.  For example, a slew up time of 0.25 would prevent the throttle from advancing from minimum to maximum faster than 0.25 seconds. 

_**Other Related Changes**_:
While Leonard and I were at it, there were a number of things in the existing motor control library that needed to be rearranged and reorganized.  So we took care of all that while we were at it.  In short, the linearization, slew rate, and conversion to PWM were broken out into separate functions, rather than being all crammed into one.  This allows for more flexibility of what control is applied in each spool mode.  The use of the word thrust was also changed to actuator_output for consistency.

_**Solo Info**_
A `MOT_SLEW_UP_TIME` of **0.25** and `MOT_SLEW_DN_TIME` of **0** will have _functionally and mathematically the same impact_ on motor output as the hard coded hack in the 3DR fork used on the Solo back in 2015.  As such, that is what will be set in the Solo parameter files.  Extensive PID tuning has been completed with these slew parameter settings, with multiple users successfully flight testing.  The PID tuning parameters have been updated in the Solo's parameter files.

Pedals2Paddles commented:
Here is some log output showing this working.  Visually, I can't see a big difference between no slew, and the 0.275 setting @lthall recommended for the Solo.  But, you can hear the difference when you punch the throttle sitting on my desk.  

**Motor PWM with `MOT_SLEW_TIME` disabled :** (slew code is completely bypassed)
![pwm no slew](https://user-images.githubusercontent.com/8234496/49310001-12651700-f4ab-11e8-9a51-5df1917aae42.png)

**Motor PWM with `MOT_SLEW_TIME` set for 0.275** as recommended by Leonard for the Solo
![pwm slew 275](https://user-images.githubusercontent.com/8234496/49310024-21e46000-f4ab-11e8-947c-b6e14c5a985e.png)

**Motor PWM with `MOT_SLEW_TIME` set for 0.999** for demonstration purposes:
![pwm slew max](https://user-images.githubusercontent.com/8234496/49310017-1e50d900-f4ab-11e8-8071-611661e1d999.png)

I would like to produce a similar graph from the old solo firmware to compare.  But that FW for some reason is producing corrupt DF logs that grow in size but have nothing to show.  If someone can make this DF log work, that would be great:
[old_solo.zip](https://github.com/ArduPilot/ardupilot/files/2634734/old_solo.zip)

tridge commented:
well done!
As I mentioned on mumble, my key concerns are:

 - what happens when we ask for zero throttle, for example when disarmed. Perhaps it should go straight to zero, and not decay slowly?
 - will some of the old thrust be used when starting a new flight, as the history isn't cleared

tridge commented on libraries/AP_Motors/AP_MotorsMulticopter.h:
```diff
@@ -192,4 +194,6 @@ class AP_MotorsMulticopter : public AP_Motors {
     // vehicle supplied callback for thrust compensation. Used for tiltrotors and tiltwings
     thrust_compensation_fn_t _thrust_compensation_callback;
+    
+    float _thrust_last[AP_MOTORS_MAX_NUM_MOTORS] = { 0 };
```
don't need to zero here

Pedals2Paddles commented:
Updated and squashed.  The safety matter discussed with Tridge has been addressed.  When `SPOOL_MODE` is in shutdown mode, the `_last_thrust[]` array variable is reset to zero.  Also, when the `SPOOL_MODE` is not in throttle_unlimited mode, the slew rate is bypassed.  As such, it can never start back up at a previously high throttle.  And it can never be slowly reducing power when the operator cuts the throttle to minimum.

Pedals2Paddles commented:
MavExplorer is opening the old 3DR version logs.  No other application to web based log analyzer is working on them. If you convert to a .log text file, everything looks to be normal, so I don't know what's wrong with it.  I even wiped the cube out with ArduPlane, then reloaded the old version of copter with no change.  So I'm unable to really analyze the old version's log in the same manner.

peterbarker commented:
On Wed, 5 Dec 2018, Matt wrote:

> MavExplorer is opening the old 3DR version logs. No other application to web based log analyzer is working on them. If you convert to a .log text file,
> everything looks to be normal, so I don't know what's wrong with it. I even wiped the cube out with ArduPlane, then reloaded the old version of copter
> with no change. So I'm unable to really analyze the old version's log in the same manner.

It's possible that every other tool you've tried assumes that the RC 
output channel message has at least 8 channel fields in it.  3DR trimmed 
that back to 4 fields IIRC.

Pedals2Paddles commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -362,10 +371,37 @@ float AP_MotorsMulticopter::get_compensation_gain() const
+    thrust_in = apply_thrust_curve_and_volt_scaling(thrust_in);
+
+    // Apply slew limit if MOT_SLEW_TIME is enabled, and spool mode is THROTTLE_UNLIMITED
+    if (is_positive(_slew_time) && _spool_mode == THROTTLE_UNLIMITED) {
```
My logic here was to make sure the slew rate wasn't impacting the other modes where we want the motors to react immediately, such stopping.  Is there a better way to do that?

Pedals2Paddles commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -362,10 +371,37 @@ float AP_MotorsMulticopter::get_compensation_gain() const
+    }
+
+    // Save the last thrust value for comparison in the next loop
+    _thrust_last[mot_num] = thrust_in;
```
For the purpose of reading the code, _last makes it clear this is the prior output being used for comparison.  At least for me, I would get lost easily if wasn't that clear.  But I'm removing all the thrusts and making them actuator_output now.

Pedals2Paddles commented:
As requested, I changed all the use of thrust to output or actuator_output in the code and comments.

Pedals2Paddles commented:
After discussing with Leonard yesterday evening, we concluded this would be a good time to break out some of the math currently all in one function into separate functions.  This will give better control over what math takes places under what circumstances.  And will also make future changes to the motor controls a lot easier.  This needs to be done with or without the slew rate.  So it's basically a matter of "while we're at it, let's get this done"

So I split out the functions and reworked everything to use them all.  I think it came out decent.  It even compiled.  However there is still something I'm missing, since the motors aren't spinning beyond minimum when increasing the throttle.  Still needs some troubleshooting, and it was time for bed.

Pedals2Paddles commented:
_Update 12/6/2018 (also included in PR description)_

**New parameters** `MOT_SLEW_UP_TIME` and `MOT_SLEW_DN_TIME` have been added. The default is 0, which applies no slew limit and makes no change to existing behavior. Valid values are 0 to 0.5 seconds. This controls the minimum amount of time that the output can go from minimum to maximum. For example, a slew up time of 0.25 would prevent the throttle from advancing from minimum to maximum faster than 0.25 seconds.

**Other Related Changes:**
While Leonard and I were at it, there were a number of things in the existing motor control library that needed to be rearranged and reorganized. So we took care of all that while we were at it. In short, the linearization, slew rate, and conversion to PWM were broken out into separate functions, rather than being all crammed into one. This allows for more flexibility of what control is applied in each spool mode. The use of the word thrust was also changed to actuator_output for consistency.

Pedals2Paddles commented:
Also, Leonard and I went through the original slew limit in the old firmware to work out exactly what settings would be appropriate in this new implementation.  The old solo slew was a maximum increase of 6 PWM steps per loop at 400hz. No slew was applied for decrease.  Taking into account the minimum and maximum limits of the output, that amounts to 0.25 seconds of slew time.

So for a solo, we'll start off testing with `MOT_SLEW_UP_TIME` set for 0.25 and `MOT_SLEW_DN_TIME` set for 0.

Pedals2Paddles commented:
Here is a DF log with all the high rate and PID logging turned on.  This is Copter 3.7-dev with this slew rate commit applied.  I have gotten the PIDs to a point where it is flyable and pretty stable.  I think it still needs some additional fine tuning.  Especially as it relates to overshooting when transitioning from fast flight to level.

Over the course of several flights on two batteries, I repeatedly punched the throttle in stabilize while descending so it has to really dig itself out of a hole.  So far, so good.

[solo_test.zip](https://github.com/ArduPilot/ardupilot/files/2658964/solo_test.zip)

lthall commented:
I think this is good to go.

Pedals2Paddles commented:
Multiple users have been test flying this all week with both stock/black and green cubes.  Flying great, no lawn darts!

tridge commented:
pending review by randy

rmackay9 commented:
great, nicely done.  I'll review this week for sure

rmackay9 commented:
@Pedals2Paddles, sorry for the delay on merging this, could you rebase on master?  It's top of my queue now.

rmackay9 commented on libraries/AP_Motors/AP_MotorsMulticopter.h:
```diff
@@ -192,4 +201,6 @@ class AP_MotorsMulticopter : public AP_Motors {
     // vehicle supplied callback for thrust compensation. Used for tiltrotors and tiltwings
     thrust_compensation_fn_t _thrust_compensation_callback;
+    
+    float _actuator[AP_MOTORS_MAX_NUM_MOTORS];
```
it would be good to add a comment re what's held in this array

rmackay9 commented on libraries/AP_Motors/AP_MotorsMulticopter.h:
```diff
@@ -126,11 +127,17 @@ class AP_MotorsMulticopter : public AP_Motors {
+    int16_t             output_to_pwm(float _actuator_output);
+    
+    // converts desired thrust to linearized actuator output in a range of 0~1
+    float               thrust_to_actuator(float _actuator_output);
```
Looks like the argument (i.e. _actuator_output) is mismatched between the .h and .cpp.  Also normally we reserve the underscore at the front for internal variables so let's remove the underscore if possible.

rmackay9 commented on libraries/AP_Motors/AP_MotorsMulticopter.h:
```diff
@@ -126,11 +127,17 @@ class AP_MotorsMulticopter : public AP_Motors {
-    // calculate spin up to pwm range
-    int16_t             calc_spin_up_to_pwm() const;
+    // gradually increase actuator output maximum limit
+    float               actuator_spin_up() const;
```
I wonder if we could come up with a more easily understood name and/or comment for the "actuator_spin_up()" function.  maybe it's "actuator_spin_up_to_ground_idle()".  That's a bit long but "spin_up" alone doesn't really clarify that it's only spinning up to ground-idle.

Pedals2Paddles commented:
This will take me a while to rebase now since another PR that was submitted after this one was merged first.  I have to redo a bunch of stuff related to differential thrust that didn't exist before.

Pedals2Paddles commented:
Actually, looks like a bunch of other stuff was changed too so this is going to be a lot of reinventiting the wheel to rebase.

rmackay9 commented:
Merged, thanks!

Pedals2Paddles commented:
Wooo!  Thank you everyone that helped with this (@lthall @tridge @rmackay9)!  The Solo can now fly 3.7 master safely!

fnoop commented:
As a side benefit, I think this should help considerably with ESC desyncs.  As I understand it, a lot of desyncs happen when using larger motors and larger props are used with ESC firmware that is designed for fast throttle punch (ie simonk and blheli), and fast active throttle stop (active braking).  One of the main tests for desyncs is to push the throttle from zero to max as fast as possible, and there are ESC firmware settings that reduce the max throttle acceleration - eg. in blheli there are startup/rampup power and maximum accleration settings, and these are important to set to stop desyncs on larger propulsion sets.  A significant portion of 'my multicopter dropped out of the sky' forum issues are desyncs, and it can be quite difficult to find/program ESCs to deal with this.

So it's great to be able to now 'fix'/mitigate this from the flight controller end, and could be very valuable to add this to the wiki for initial setup, and anywhere else that deals with problems/troubleshooting this kind of problem.

What would be the downsides of changing the default to a higher value?

Pedals2Paddles commented:
Leonard said the same thing when were collaborating on this. So Solos are probably going to be the first and most common users, it definitely has application beyond that.  

The default is 0 (disabled) for both up and down.  The solo uses .25 for throttling up, and 0 for throttling down.  The hard coded max is .5.  Going beyond .5 I think would make tuning and control rather difficult.
